### DISCUSSION TITLE
Let's cooperate with Apple Silicon support

### DISCUSSION THREAD

**Discussion by simnalamburt** [heyvito/podman-macos#2] [Show and tell]
Hi @heyvito! I have [patched and packaged qemu and podman](https://github.com/simnalamburt/homebrew-x/) to make it work with Apple Silicon, and now I'm looking for the way to integrate it with podman-macos.

This is how to use patched version of podman in the CLI:
```bash
brew install simnalamburt/x/podman-apple-silicon

# Example
podman machine init
podman machine start
podman run -p 8080:80 docker.io/nginx
```

I tried to use it with podman-macos and this is the issues I get:

1.  podman-macos do not search podman binary in `/opt/homebrew/bin`

    You can workaround this issue by `sudo ln -s /opt/homebrew/bin/podman /usr/local/bin/podman`, but it would be better if podman-macos checks the `/opt/homebrew/bin`. Maybe we can fix this by appending `/opt/homebrew/bin` at the end of the result of [`PMManager::paths`](https://github.com/heyvito/podman-macos/blob/7f3b8933d6eed5db438dfc659e5e3b7812580a1c/Podman/Interop/PMManager.m#L95).

2.  podman-macos successfully reads states from podman but fails to write.

    After you deal with locating the podman binary, you'll find podman-macos successfully reads information from podman.

    <img width="437" alt="스크린샷 2021-09-05 05 56 52" src="https://user-images.githubusercontent.com/4435445/132107866-9e269f50-e0ee-48fa-8206-f7d586984e73.png">

    But unfortunately any operation that tries to write something fails. For example, podman-macos is unable to stop podman VM.

    <img width="328" alt="스크린샷 2021-09-05 05 55 33" src="https://user-images.githubusercontent.com/4435445/132107843-8252134e-b026-49bd-a0f3-375c33b0b75a.png">

I'm wiling to fix these issues but I don't know how to build or debug this project. Is there any chance for us to cooperate with these issues?

**Comment by heyvito** [comment]
Hi there @simnalamburt! Thank you for your interest!

I think that it would be ideal to bundle qemu, the VM base image, Podman itself (and perhaps Podman-compose). Could you share how you built the aarm64 version you're running in your system? We could perhaps write a script to be able to consistently create the binary, the image, and so.

You may be able to see what went wrong looking at Console.app's output, and filtering by Podman using its search field. I may have forgotten to emit log entries when something goes very wrong (like the case you mentioned), but adding those logs is quite simple, and I could also use a new release to see if Sparkle is able to auto-update the application in your system. :P

**Comment by simnalamburt** [comment]
@heyvito I built aarch64 version of podman and forked version of qemu with following commands:

```bash
brew install simnalamburt/x/podman-apple-silicon
```

and the definition of that formula is located at https://github.com/simnalamburt/homebrew-x repo:

- [podman-apple-silicon](https://github.com/simnalamburt/homebrew-x/blob/main/Formula/podman-apple-silicon.rb)
- [qemu-hvf](https://github.com/simnalamburt/homebrew-x/blob/main/Formula/qemu-hvf.rb)

And you'll be able to find pre-built binaries at github releases page:

- https://github.com/simnalamburt/homebrew-x/releases/tag/podman-apple-silicon-4.0.0-dev%2B20210905
- https://github.com/simnalamburt/homebrew-x/releases/tag/qemu-hvf-6.1.0-patched.0

And thanks for the note for the debugging. I'll try that method.

**Reply by sammcj** [reply]
Hey,

Just wondering if this is going to be merged in?

**Comment by simnalamburt** [comment]
Almost all patches has been merged into upstream main branch but AFAIK not
released yet.

2021년 10월 26일 (화) 08:27, Sam ***@***.***>님이 작성:

**Comment by mikebell90** [comment]
@simnalamburt is it not integrated and released? The podman blog implies it was and I successfully installed and used podman on a brand new m1 using just brew. To get x64 running I had to do the hack here https://edofic.com/posts/2021-09-12-podman-m1-amd64/
but that worked ok too.

**Comment by simnalamburt** [comment]
Oh really? Then the new release which includes M1 support must have been
released.

I tried to merge all custom patches that I have used in this fork to the
upstream, and those contain enhanced versions of the patch. So if they have
released the newer version of podman with M1 support, you'd be better to
use them. Good news!

Update: To clarify, https://github.com/simnalamburt/podman/issues/5#issue-993566828 these are the patches that is required for podman's M1 support and looks like all of them has been merged and released as podman v3.4.0

CC @sammcj
